pandas/tests/test_categorical.py

```diff
@@ -1673,9 +1673,11 @@ def test_unique(self):
         exp_cat = Categorical(["b", np.nan, "a"], categories=["b", "a"])
         tm.assert_categorical_equal(res, exp_cat)
 
-        # GH 18051 unique()._codes should be writeable
-        cat_nan = Categorical([np.nan])
-        assert cat_nan.unique()._codes.flags.writeable
+        # GH 18051
+        cat = Categorical([np.nan])
+        res = cat.unique()
+        exp_cat = Categorical([np.nan], categories=[])
+        tm.assert_categorical_equal(res, exp_cat)
```

Comment:
@jreback , do you mean like this? (This is not failing on `0.20.3` though)

Reply:
There should be a test for

```python
>>> assert pd.Series(pd.Categorical([np.nan])).nunique() == 0
```
Note this is a method on ``Series``, not ``Categorical``.

Reply:
@topper-123  I have added that test in [https://github.com/pandas-dev/pandas/pull/18279/files#diff-ed4f442894a2f521dfac3193a3a8d8a0R2185](L2185) below.

Reply:
That looks great IMO.

Also, you mention above that ``Categorical([np.nan]).unique()`` doesn't fail in 0.21.0, but ``pd.Series(pd.Categorical([np.nan])).unique()`` does fail. So if you could add a test for that as well in the test for series, then IMO the PR is good (but I'll let @jreback make the final call on that).